runtime/router.go

```diff
@@ -108,72 +40,50 @@ type HandlerFn func(
 	*ServerHTTPResponse,
 )
 
-// EndpointMetrics contains pre allocated metrics structures
-// These are pre-allocated to cache tags maps and for performance
-type EndpointMetrics struct {
-	requestRecvd   tally.Counter
-	requestLatency tally.Timer
-	statusCodes    map[int]tally.Counter
-}
-
 // RouterEndpoint struct represents an endpoint that can be registered
 // into the router itself.
 type RouterEndpoint struct {
 	EndpointName string
 	HandlerName  string
 	HandlerFn    HandlerFn
-
-	metrics EndpointMetrics
-	gateway *Gateway
+	Logger       *zap.Logger
+	metrics      *InboundHTTPMetrics
 }
 
 // NewRouterEndpoint creates an endpoint with all the necessary data
 func NewRouterEndpoint(
-	gateway *Gateway,
 	endpointName string,
 	handlerName string,
 	handler HandlerFn,
+	logger *zap.Logger,
+	scope tally.Scope,
 ) *RouterEndpoint {
-	endpointTags := map[string]string{
-		"endpoint": endpointName,
-		"handler":  handlerName,
-	}
-	endpointScope := gateway.AllHostScope.Tagged(endpointTags)
-	requestRecvd := endpointScope.Counter("inbound.calls.recvd")
-	requestLatency := endpointScope.Timer("inbound.calls.latency")
-	statusCodes := make(map[int]tally.Counter, len(knownStatusCodes))
-
-	for _, statusCode := range knownStatusCodes {
-		metricName := "inbound.calls.status." + strconv.Itoa(statusCode)
-		statusCodes[statusCode] = endpointScope.Counter(metricName)
-	}
-
 	return &RouterEndpoint{
 		EndpointName: endpointName,
 		HandlerName:  handlerName,
 		HandlerFn:    handler,
-		gateway:      gateway,
-
-		metrics: EndpointMetrics{
-			requestRecvd:   requestRecvd,
-			statusCodes:    statusCodes,
-			requestLatency: requestLatency,
-		},
+		Logger:       logger,
+		metrics:      NewInboundHTTPMetrics(scope, logger),
 	}
 }
 
 // HandleRequest is called by the router and starts the request
 func (endpoint *RouterEndpoint) HandleRequest(
-	w http.ResponseWriter, r *http.Request, params httprouter.Params,
+	w http.ResponseWriter,
+	r *http.Request,
+	params httprouter.Params,
 ) {
-	reqFields := logRequestFields(r)
+	startTime := time.Now()
+	var req *ServerHTTPRequest
 	var resFields []zapcore.Field
+	reqFields := logRequestFields(r)
 
 	defer func() {
-		writeLogs(endpoint.gateway.Logger, reqFields, resFields)
+		writeLogs(req.Logger, reqFields, resFields)
+		endpoint.metrics.CollectMetrics(startTime, req.res.StatusCode)
```

Comment:
We do not want to do metrics like this. Makes the code too difficult to follow. please revert to emitting metrics inline.